src/tactic/lift.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2019 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import tactic.rcases
+
+/-!
+# lift tactic
+
+This file defines the lift tactic, allowing the user to lift elements from one type to another
+under a specified condition.
+
+## Tags
+
+lift, tactic
+-/
+
+universe variables u v
+
+run_cmd mk_simp_attr `can_lift
+
+/-- A class specifying that you can lift elements from `α` to `β` assuming `cond` is true.
+  Used by the tactic `lift`. -/
+class can_lift (α : Type u) (β : Type v) : Type (max u v) :=
+(coe : β → α)
+(cond : α → Prop)
+(prf : ∀(x : α), cond x → ∃(y : β), coe y = x)
+
+@[can_lift] instance : can_lift ℤ ℕ :=
+⟨coe, λ n, 0 ≤ n, λ n hn, ⟨n.nat_abs, int.nat_abs_of_nonneg hn⟩⟩
+
+namespace tactic
+
+meta def lift (p : pexpr) (t : pexpr) (h : option pexpr) (n : list name) : tactic unit :=
+do
+  let dsimp := interactive.dsimp tt [] [`can_lift],
+  e ← i_to_expr p,
+  old_tp ← infer_type e,
+  new_tp ← i_to_expr t,
+  inst_type ← mk_app ``can_lift [old_tp, new_tp],
+  inst ← mk_instance inst_type <|>
+    pformat!"Failed to find a lift from {old_tp} to {new_tp}. Provide an instance of\n  {inst_type}"
+    >>= fail,
+  prf ← if h_some : h.is_some then
+    (do prf ← i_to_expr (option.get h_some), prf_ty ← infer_type prf,
+    expected_prf_ty ← mk_app `can_lift.cond [old_tp, new_tp, inst, e],
+    unify prf_ty expected_prf_ty <|>
+      (do expected_prf_ty2 ← expected_prf_ty.dsimp {} tt [`can_lift],
+        pformat!"lift tactic failed. The type of\n  {prf}\nis\n  {prf_ty}\nbut it is expected to be\n  {expected_prf_ty2}" >>= fail),
+    return prf)
+    else (do prf_nm ← get_unused_name,
+      prf ← mk_app `can_lift.cond [old_tp, new_tp, inst, e] >>= assert prf_nm,
+      focus1 (dsimp $ interactive.loc.ns [none]), swap, return prf),
+  let prf_nm := if prf.is_local_constant then some prf.local_pp_name else none,
+  /- We use mk_mapp to apply `can_lift.prf` to all but one argument, and then just use expr.app
+  for the last argument. For some reason we get an error when applying mk_mapp it to all
+  arguments. -/
+  prf_ex_partial ← mk_mapp `can_lift.prf [old_tp, new_tp, inst, e],
+  let prf_ex := prf_ex_partial prf,
+  new_nm ← if n ≠ [] then return n.head
+    else if e.is_local_constant then return e.local_pp_name
+    else get_unused_name,
+  eq_nm ← if hn : 1 < n.length then return (n.nth_le 1 hn)
+    else if e.is_local_constant then return `rfl
+    else get_unused_name `h,
+  temp_nm ← get_unused_name,
+  temp_e ← note temp_nm none prf_ex,
+  dsimp (interactive.loc.ns [temp_nm]),
```

Comment:
E.g. this `dsimp` could use a comment. I wasn't sure what was happening here until I saw your answer to Simon.